# Importing Libraries

In [1]:
from __future__ import division
import numpy as np
import matplotlib.pyplot as plt 

from sklearn.metrics import accuracy_score
from sklearn.utils import shuffle
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
np.random.seed(42)

In [2]:
print(load_wine().DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

In [3]:
wine_data = load_wine()
X = wine_data.data
Y = wine_data.target.reshape((-1,1))

In [4]:
y = np.zeros((X.shape[0],3))
for i in range(len(Y)):
    y[i][Y[i]] = 1


In [5]:
X = X.T
for i in range(len(X)):
    X[i] = X[i]/np.max(X[i])
X = X.T
X.shape

(178, 13)

In [6]:
(X,y) = shuffle(X,y,random_state = 40)

In [7]:
X_train, X_test, Y_train, Y_test = train_test_split(X, y, test_size = 0.20)

In [8]:
def relu(x):
    x[x<0]=0
    return x

def softmax(x):
    return np.exp(x)/(np.sum(np.exp(x),axis=0))

def diff_relu(x):
    y = np.zeros(x.shape)
    y[x<=0] = 0
    y[x>0] = 1
    return y

def weight_init(x,y):
    return np.sqrt(2.0/(x+y))*np.random.normal(0,1,(x,y))

In [9]:
inp_dim = 13
hl1_units = 15
hl2_units = 15
out_dim = 3

W1 = weight_init(hl1_units,inp_dim)
b1 = weight_init(hl1_units,1)
W2 = weight_init(hl2_units,hl1_units)
b2 = weight_init(hl2_units,1)
W3 = weight_init(out_dim,hl2_units)
b3 = weight_init(out_dim,1)

# Fitting training data using Adam approach for adaptive learning rate

In [10]:
epochs = 290
epsilon = 0.001
delta = 1e-8
rho1 = 0.9
decay_rate = 0.999

for i in range(epochs):
    (x_train,y_train) = shuffle(X_train,Y_train,random_state = 40)
    loss = 0
    
    v = {"W1":np.zeros(W1.shape), "W2": np.zeros(W2.shape), "W3": np.zeros(W3.shape), "b1":np.zeros(b1.shape), "b2":np.zeros(b2.shape), "b3":np.zeros(b3.shape)}
    
    r = {"W1":np.zeros(W1.shape), "W2": np.zeros(W2.shape), "W3": np.zeros(W3.shape), "b1":np.zeros(b1.shape), "b2":np.zeros(b2.shape), "b3":np.zeros(b3.shape)}
    for j in range(0,len(x_train)):
        
        a1 = relu(np.matmul(W1,x_train[j]).reshape((-1,1)) + b1)
        a2 = relu(np.matmul(W2,a1).reshape((-1,1)) + b2)
        x = np.matmul(W3,a2) + b3
        a3 = softmax(x)
        loss += -np.log(a3[np.argmax(y_train[j])]) 
        
        delta3 = a3 - y_train[j].reshape(-1,1)
        delta2 = np.matmul(W3.T,delta3)*diff_relu(a2)
        delta1 = np.matmul(W2.T,delta2)*diff_relu(a1)
        
        grd_b3 = delta3
        grd_W3 = np.matmul(delta3,a2.T)
        grd_b2 = delta2
        grd_W2 = np.matmul(delta2,a1.T)
        grd_b1 = delta1
        grd_W1 = np.matmul(delta1,x_train[j].reshape((-1,1)).T)
        
        v["W1"] = rho1*v["W1"] + (1-rho1)*grd_W1
        v["b1"] = rho1*v["b1"] + (1-rho1)*grd_b1
        v["W2"] = rho1*v["W2"] + (1-rho1)*grd_W2
        v["b2"] = rho1*v["b2"] + (1-rho1)*grd_b2
        v["W3"] = rho1*v["W3"] + (1-rho1)*grd_W3
        v["b3"] = rho1*v["b3"] + (1-rho1)*grd_b3
        
        r["W1"] = decay_rate*r["W1"]+(1-decay_rate)*(grd_W1*grd_W1)
        r["b1"] = decay_rate*r["b1"]+(1-decay_rate)*(grd_b1*grd_b1)
        r["W2"] = decay_rate*r["W2"]+(1-decay_rate)*(grd_W2*grd_W2)
        r["b2"] = decay_rate*r["b2"]+(1-decay_rate)*(grd_b2*grd_b2)
        r["W3"] = decay_rate*r["W3"]+(1-decay_rate)*(grd_W3*grd_W3)
        r["b3"] = decay_rate*r["b3"]+(1-decay_rate)*(grd_b3*grd_b3)
        
        W1 += -epsilon*(v["W1"]/(1-rho1**(j+1)))/np.sqrt(delta+(r["W1"])/(1-decay_rate**(j+1)))
        b1 += -epsilon*(v["b1"]/(1-rho1**(j+1)))/np.sqrt(delta+(r["b1"])/(1-decay_rate**(j+1)))
        W2 += -epsilon*(v["W2"]/(1-rho1**(j+1)))/np.sqrt(delta+(r["W2"])/(1-decay_rate**(j+1)))
        b2 += -epsilon*(v["b2"]/(1-rho1**(j+1)))/np.sqrt(delta+(r["b2"])/(1-decay_rate**(j+1)))
        W3 += -epsilon*(v["W3"]/(1-rho1**(j+1)))/np.sqrt(delta+(r["W3"])/(1-decay_rate**(j+1)))
        b3 += -epsilon*(v["b3"]/(1-rho1**(j+1)))/np.sqrt(delta+(r["b3"])/(1-decay_rate**(j+1)))
        
    if i%10 == 0:
        print (str(i) + ":" + " loss =" + str(loss/len(x_train)))

0: loss =[1.10329907]
10: loss =[0.34852205]
20: loss =[0.18121837]
30: loss =[0.1222222]
40: loss =[0.08993276]
50: loss =[0.06847813]
60: loss =[0.05585418]
70: loss =[0.04489173]
80: loss =[0.03662108]
90: loss =[0.02976965]
100: loss =[0.02517135]
110: loss =[0.02158477]
120: loss =[0.01995091]
130: loss =[0.01820445]
140: loss =[0.01627512]
150: loss =[0.01485278]
160: loss =[0.0137053]
170: loss =[0.01282053]
180: loss =[0.00938519]
190: loss =[0.01054907]
200: loss =[0.01340968]
210: loss =[0.00673754]
220: loss =[0.00416276]
230: loss =[0.00623627]
240: loss =[0.02368546]
250: loss =[0.00315239]
260: loss =[0.00072088]
270: loss =[0.00336858]
280: loss =[0.00047611]


# Model Accuracy

In [11]:
y_predicted = list()
y_actual = list()
for j in range(0,len(X_test)):
    a1 = relu(np.matmul(W1,X_test[j]).reshape((-1,1)) + b1)
    a2 = relu(np.matmul(W2,a1).reshape((-1,1)) + b2)
    x = np.matmul(W3,a2) + b3
    out = softmax(x)
    y_predicted.append(np.argmax(out))
    y_actual.append(np.argwhere(Y_test[j] == 1)[0][0])
print ("actual out:"+str(y_actual))
print ("predic out:"+str(y_predicted))

actual out:[0, 1, 1, 0, 0, 2, 1, 0, 1, 1, 2, 0, 2, 1, 2, 0, 2, 0, 0, 1, 1, 1, 1, 2, 1, 0, 1, 1, 2, 0, 1, 1, 1, 0, 0, 2]
predic out:[0, 1, 1, 0, 0, 2, 1, 0, 1, 1, 2, 0, 2, 1, 2, 0, 2, 0, 0, 1, 0, 1, 1, 2, 1, 0, 1, 0, 2, 0, 1, 1, 1, 0, 0, 2]


# Comparision with RMSProp and Adagrad approach

### Loss obtained with 1000 epochs with Adagrad is nearly 0.02452002 and in RMSProp approach loss of 0.00336388 is achieved in 300 epochs where as in Adam approach a loss of 0.00047611 is achieved in 290 epochs which clearly shows that model converges at a faster rate using RMSProp approach than that with RMSProp and Adagrad approach